src/renderer/object/instanced_model.rs

```diff
@@ -49,53 +52,173 @@ impl<M: ForwardMaterial> InstancedModel<M> {
             instance_buffer3: InstanceBuffer::new(context)?,
+            instance_buffer4: InstanceBuffer::new(context)?,
             aabb,
             aabb_local: aabb.clone(),
             transformation: Mat4::identity(),
             transformations: transformations.to_vec(),
+            subtexture: None,
+            subtextures: None,
             material,
         };
-        model.update_transformations(transformations);
-        model.update_aabb();
+        model.set_transformations(transformations);
         Ok(model)
     }
 
     ///
-    /// Updates the transformations applied to each model instance before they are rendered.
-    /// The model is rendered in as many instances as there are transformation matrices.
+    /// Use [set_transformations()](InstancedModel::set_transformations()) instead.
     ///
+    #[deprecated]
     pub fn update_transformations(&mut self, transformations: &[Mat4]) {
+        self.set_transformations(&transformations);
+    }
+
+    ///
+    /// Store the transformations applied to each model instance before they are rendered.
+    /// The model is rendered in as many instances as there are transformation matrices.
+    ///
+    pub fn set_transformations(&mut self, transformations: &[Mat4]) {
         self.transformations = transformations.to_vec();
-        self.instance_count = transformations.len() as u32;
+        self.update_buffers();
+    }
+
+    ///
+    /// Store the transformations applied to each model instance before they are rendered.
+    /// The model is rendered in as many instances as there are transformation matrices.
+    /// Must (eventually) call [update_buffers()](InstancedModel::update_buffers()) afterward.
+    ///
+    pub fn add_transformation(&mut self, transformation: &Mat4) {
+        self.transformations.push(*transformation);
+    }
+
+    ///
+    /// Clear transformations array.
+    ///
+    pub fn clear_transformations(&mut self) {
+        self.transformations = Vec::new();
+        self.update_buffers();
+    }
+
+    ///
+    /// Update instance transform and uv buffers and aabb.
+    ///
+    pub fn update_buffers(&mut self) {
+        self.instance_count = self.transformations.len() as u32;
+        let transformations = self.transformations.as_slice();
         let mut row1 = Vec::new();
         let mut row2 = Vec::new();
         let mut row3 = Vec::new();
-        for transform in transformations {
-            row1.push(transform.x.x);
-            row1.push(transform.y.x);
-            row1.push(transform.z.x);
-            row1.push(transform.w.x);
-
-            row2.push(transform.x.y);
-            row2.push(transform.y.y);
-            row2.push(transform.z.y);
-            row2.push(transform.w.y);
-
-            row3.push(transform.x.z);
-            row3.push(transform.y.z);
-            row3.push(transform.z.z);
-            row3.push(transform.w.z);
+        let mut subt = Vec::new();
+        if self.transformations.len() > 0 {
```

Comment:
The transformation is a global transformation applied to all instances and each local transformation in the list of transformations are applied to each instance. We apply both the global and the local transformation so you should just send the list of transformations as it did before.